### PR TITLE
Disable foreign distributed tables

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: Disables distributed&reference foreign tables

Disables `create_distributed_table` and `create_reference_table` for foreign tables. Foreign tables can now only be added to metadata. Errors out when trying to distribute foreign tables.

We don't want to break users that have distributed&reference foreign tables already. With this PR, we log a WARNING message for them, when they use an existing distributed foreign table. This check is done in distributed planner.

Includes a lot of changes in regression test output, for the existing distributed foreign table test.
Added some test to show the error when trying to distribute a foreign table.
Can't add regression test for the warning, as it's not a case we can see for now. The warning is only for the users that have distributed foreign tables already.

fixes: #5604

### CODE REVIEW

**Review by onderkalaci — COMMENTED** [review]
I think the question @marcocitus had earlier was that what about existing foreign distributed tables that were created on 10.2 or earlier. 

Can you still (a) query those (b) run alter/rename/DDLs etc (c) drop them

Also, when you do (a) and (b), should we warn users?

**Comment by onderkalaci** [comment]
> Also, when you do (a) and (b), should we warn users?

Well, that might be too much. Maybe warn once per session or such on these occasions.

**Comment by agedemenli** [comment]
> I think the question @marcocitus had earlier was that what about existing foreign distributed tables that were created on 10.2 or earlier.
> 
> Can you still (a) query those (b) run alter/rename/DDLs etc (c) drop them
> 
> Also, when you do (a) and (b), should we warn users?

Yes, all of them are still allowed. I'll add a warning for such cases.

File: src/backend/distributed/planner/distributed_planner.c [hunk 1]
```diff
@@ -336,6 +337,7 @@ ListContainsDistributedTableRTE(List *rangeTableList)
 
 		if (IsCitusTable(rangeTableEntry->relid))
 		{
+			WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:340]: I'm really not convinced that calling `WarnUnsupportedIfForeignDistributedTable` inside `ListContainsDistributedTableRTE` is a good approach. 

What about:
```C
	else if (CitusHasBeenLoaded())
	{
		needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList, maybeHasForeignTable);
		if (needsDistributedPlanning)
		{
			fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);
			
			if (maybeHasForeignTable)
			{
			
                 WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
			}
		
		}
	}
```

Comment (agedemenli) [reply src/backend/distributed/planner/distributed_planner.c:340]: We actually need a relationId to pass to `WarnUnsupportedIfForeignDistributedTable`, to verify the relation is a distributed foreign table. But here we don't have a relationId, instead we have a list that might contain a relation that is a distributed foreign table.

Besides, there is one more place that we call `ListContainsDistributedTableRTE`, that has nothing to do with foreign tables, so passing such a parameter doesn't make sense in that place.

We can consider iterating the list once more, to find out if it contains any distributed foreign table; but this option adds some unnecessary complexity. Maybe we can avoid this complexity by doing the iteration only when the variable `DistributedForeignTableWarningPrompted` is `false` ?

Something like:
```c
	else if (CitusHasBeenLoaded())
	{
		needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList);
		if (needsDistributedPlanning)
		{
			fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);

			WarnIfListHasForeignDistributedTable(rangeTableList);
		}
	}
```
and 
```c
static void
WarnIfListHasForeignDistributedTable(List *rangeTableList)
{
	RangeTblEntry *rangeTableEntry = NULL;
	foreach_ptr(rangeTableEntry, rangeTableList)
	{
		if (DistributedForeignTableWarningPrompted)
		{
			return;
		}

		WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
	}
}
```

Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c:340]: > Besides, there is one more place that we call ListContainsDistributedTableRTE, that has nothing to do with foreign tables, so passing such a parameter doesn't make sense in that place.

Similarly, calling `WarnUnsupportedIfForeignDistributedTable` doesn't make sense :)

 Can't we re-iterate the list only when we know there is a foreign table? In majority of the cases `DistributedForeignTableWarningPrompted ` will be false, hence we'd re-iterate unnecessarily with the above suggestion

File: src/backend/distributed/commands/table.c [hunk 2]
```diff
@@ -3403,3 +3405,44 @@ MakeNameListFromRangeVar(const RangeVar *rel)
...
 }
+
+
+/*
+ * WarnUnsupportedIfForeignDistributedTable gets a relationId and logs a WARNING
+ * if the given relation is a distributed foreign table.
+ * We do that because now we only support Citus Local Tables for foreign tables.
+ */
+void
+WarnUnsupportedIfForeignDistributedTable(Oid relationId)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:3416]: do we really need this function now? Can't we inline into `WarnIfListHasForeignDistributedTable`?  It seems confusing to have an external function with this name. 

And, when we do that, we should probably be able to remove `DistributedForeignTableWarningPrompted` check from this part?

File: src/backend/distributed/commands/table.c [hunk 3]
```diff
@@ -3403,3 +3405,44 @@ MakeNameListFromRangeVar(const RangeVar *rel)
...
+ * WarnUnsupportedIfForeignDistributedTable for each relation, which logs
+ * a WARNING message if it's a distributed foreign table.
+ */
+void
+WarnIfListHasForeignDistributedTable(List *rangeTableList)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		if (DistributedForeignTableWarningPrompted)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:3441]: why not this check is done before the loop? 
Using `DistributedForeignTableWarningPrompted` in two different functions make it pretty hard to follow. Can you make sure we only use it once? 

And, can't we define that as a function level static variable?

File: src/backend/distributed/planner/distributed_planner.c [hunk 4]
```diff
@@ -336,6 +351,11 @@ ListContainsDistributedTableRTE(List *rangeTableList)
 
 		if (IsCitusTable(rangeTableEntry->relid))
 		{
+			if (IsForeignTable(rangeTableEntry->relid))
+			{
+				*maybeHasForeignDistributedTable = true;
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:356]: what if `maybeHasForeignDistributedTable == NULL`? we should be robust to that

File: src/backend/distributed/commands/table.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:3441]: See function scoped static variables: https://www.geeksforgeeks.org/static-variables-in-c/

File: src/backend/distributed/commands/table.c [hunk 5]
```diff
@@ -3403,3 +3403,37 @@ MakeNameListFromRangeVar(const RangeVar *rel)
...
 }
+
+
+/*
+ * WarnIfListHasForeignDistributedTable iterates the given list and logs a WARNING
+ * if the given relation is a distributed foreign table.
+ * We do that because now we only support Citus Local Tables for foreign tables.
+ */
+void
+WarnIfListHasForeignDistributedTable(List *rangeTableList)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:3414]: minor: probably better to move where it is used, and make it a static function

File: src/backend/distributed/commands/table.c [hunk 6]
```diff
@@ -3403,3 +3403,37 @@ MakeNameListFromRangeVar(const RangeVar *rel)
...
+{
+	static bool DistributedForeignTableWarningPrompted = false;
+
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		if (DistributedForeignTableWarningPrompted)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:3421]: minor: might be good to move this check out of the loop

**Review by onderkalaci — APPROVED** [review]
